File: PHP/src/OneLang/One/Transforms/ResolveGenericTypeIdentifiers.php
<?php

namespace OneLang\One\Transforms\ResolveGenericTypeIdentifiers;

use OneLang\One\AstTransformer\AstTransformer;
use OneLang\One\Ast\AstTypes\UnresolvedType;
use OneLang\One\Ast\AstTypes\GenericsType;
use OneLang\One\Ast\Types\Class_;
use OneLang\One\Ast\Types\Method;
use OneLang\One\Ast\Interfaces\IType;

class ResolveGenericTypeIdentifiers extends AstTransformer {
    function __construct() {
        parent::__construct("ResolveGenericTypeIdentifiers");
        
    }
    
    protected function visitType($type) {
        parent::visitType($type);
        
        //console.log(type && type.constructor.name, JSON.stringify(type));
        if ($type instanceof UnresolvedType && (($this->currentInterface instanceof Class_ && in_array($type->typeName, $this->currentInterface->typeArguments)) || ($this->currentMethod instanceof Method && in_array($type->typeName, $this->currentMethod->typeArguments))))
            return new GenericsType($type->typeName);
        
        return $type;
    }
}

File: Python/OneLang/One/Transforms/ResolveGenericTypeIdentifiers.py
from onelang_core import *
import OneLang.One.AstTransformer as astTrans
import OneLang.One.Ast.AstTypes as astTypes
import OneLang.One.Ast.Types as types
import OneLang.One.Ast.Interfaces as ints

class ResolveGenericTypeIdentifiers(astTrans.AstTransformer):
    def __init__(self):
        super().__init__("ResolveGenericTypeIdentifiers")
    
    def visit_type(self, type):
        super().visit_type(type)
        
        #console.log(type && type.constructor.name, JSON.stringify(type));
        if isinstance(type, astTypes.UnresolvedType) and ((isinstance(self.current_interface, types.Class) and type.type_name in self.current_interface.type_arguments) or (isinstance(self.current_method, types.Method) and type.type_name in self.current_method.type_arguments)):
            return astTypes.GenericsType(type.type_name)
        
        return type